presentations to OCF, directory structure for liaisons (proposed/actual)

mmccool opened the pull request:
This includes the presentation on the W3C WoT I gave on 2016-11-10 at the OCF AGM in Taipei, a slightly updated version of the same presentation fixing some issues, and a proposed structure for managing liaison presentations and documents (not only for OCF).

h0ru5 commented:
Good idea to have liason material here!

benfrancis commented:
I noticed that this presentation claims that Mozilla is a member of the Web of Things Interest Group. As I understand it Mozilla is not currently a member of the Interest Group, it has expressed an interest in joining the Web of Things Working Group, but has made a formal [objection](https://lists.w3.org/Archives/Public/www-archive/2016Oct/0004.html) to the current proposed charter.

Also, I'm sure there's a better place for this feedback but I find the content of this presentation quite confusing. It claims that the target of the WoT Working Group is to apply "web technologies to the Internet of Things" but that doesn't seem to be reflected in the technical approach described. By trying to represent Things as abstract "software objects" which are agnostic to any programming language or protocol it is in fact quite indifferent to the web, by aiming to be agnostic as to web protocols (HTTP) and non-web protocols (MQTT/XMPP). It's also unclear to me how "each thing has a URI for its name" would work when binding to MQTT and XMPP as opposed to HTTP and CoAP.

The technical approach described in this document does not seem to be about a Web of Things but rather a semantic vocabulary and abstract scripting API for the Internet of Things which are agnostic as the web and other Internet application layers.

draggett commented:
@benfrancis The role of URIs for naming things is in respect to their descriptions, and this is independent of the protocol used to access particular devices. The descriptions include the kinds of things and their relationships, thereby forming a Web of things that can be spidered just like the Web of pages.

Mozilla used to be a member of the WoT IG, so apologies for forgetting to remove your logo from the slide with all the logos. I hope to see Mozilla rejoin given the benefits to Mozilla as a platform provider, that will accrue from the massive expansion of the market as a result of having a platform, protocol and application domain independent abstraction layer for the IoT. Netscape and subsequently Mozilla certainly benefitted from the exponential growth, the Web enabled over the last twenty years.

benfrancis commented:
@draggett I see, so Things are named with URIs, but those URIs are not necessarily URLs on the web. That would seem to reinforce my point that this is not a Web of Things proposal but rather an Internet of Things proposal.

Oh sorry, I wasn't aware that Mozilla was previously a member. Who was the representative from Mozilla and when did they leave? For what it's worth I actually requested to join the Interest Group in February but I was discouraged from doing so (internally) at the time because our work in this area wasn't sufficiently developed. I think that has changed now, which is why we are interested in joining the Working Group, but not under the currently proposed charter.

Mozilla has certainly benefited from the expansion of the *web*, but with respect I don't think it's in Mozilla's, or the web's, interest to promote a "platform, protocol and application domain independent absstraction layer for the IoT". In my opinion the currently proposed solution would actually risk harming the web and interoperability rather than improving it.

draggett commented:
@benfrancis, the URIs for things would either local to the LAN or on the Internet depending on the setup.  Sometimes you might want to have things that are published and consumed on the home wifi network without making them public on the Internet.

mmccool commented:
@benfrancis, Sorry that Mozilla was on that slide, I will edit it and upload a new version.   Of course it would be even better if we could entice Mozilla to re-join the IG so the slide would be correct again.   It would be useful to have Mozilla's perspective in the group.

Regarding URIs vs. URLs, etc, the fact is that IoT devices, for various reasons, some of them technical (eg small battery-operated devices that can't act as servers because they are only on twice a day, or because they don't have the memory to support IP, or whatever) have to use a variety of protocols outside of HTTP or CoAP.   If we restrict ourselves to just devices that use HTTP and CoAP, then it would just be a "Web of Some of the Things" which is not really useful.   Instead, we want to *connect* the current Web to the IoT and enable Web functionality for IoT devices: search, universal data transport, access to web services, and so forth.   I emphasized in particular in my talk the role of the W3C effort to create bridges between the Web and IoT devices with a mechanism to map to and from RESTful APIs for devices that are not "native" HTTP/CoAP devices.   I also personally think we should aim for a design that enables tooling for applications like search, validation, security auditing, automatic bridging, and rapid development.

I don't think the above approach is in conflict with the idea of the "Web of Things" because the goal is still to first, bridge devices into RESTful protocols like HTTP as soon as possible, and secondly, to use re-use web technologies (including semantic web technologies), and third, to enable web functionality like search.

In addition, it was my understanding was that we did not want to constrain how Thing Descriptions were accessed since while they might in many cases come over HTTP (either from a separate server, a gateway, or the device itself if it supports HTTP) they may also be provided over future DDoS-attack-resilient distributed mechanisms that have not yet been invented.   A URI certainly *includes* HTTP and CoAP though and I expect these to be common mechanisms.

Regarding the charter, we have been working on modifying it to address the concerns raised by Mozilla in your formal objection.  There are some pull requests for the charter outstanding.  They are still being discussed within the IG, but I expect we will come to a consensus soon on an updated proposal and then I would like to reach out to Mozilla specifically to discuss if we have addressed your concerns.

mmccool commented:
By the way, the "bridge" issue is why Things are "software representations of objects" since they may be proxies running in gateways that are translating from HTTP (say) to some other protocol to talk to the actual object.  You can't assume the Thing you are talking to is actually ON the device in question.  But by representing the target as well with a common metadata format (eg a Thing Description), even if the target is not using HTTP, it will be a lot easier to generate a translation.

benfrancis commented:
@mmccool 

> Of course it would be even better if we could entice Mozilla to re-join the IG so the slide would be correct again.

I would very much like that too! The concern raised back in February when I requested to join was that joining the group might imply we agreed with its current direction. I'm working very hard to promote the Web of Things approach to IoT inside Mozilla, but with the current proposed charter I'm struggling to get it taken seriously. This is my motivation for the changes I'm proposing.

> Regarding URIs vs. URLs, etc, the fact is that IoT devices, for various reasons, some of them technical (eg small battery-operated devices that can't act as servers because they are only on twice a day, or because they don't have the memory to support IP, or whatever) have to use a variety of protocols outside of HTTP or CoAP. If we restrict ourselves to just devices that use HTTP and CoAP, then it would just be a "Web of Some of the Things" which is not really useful.

Again, there's probably a better place for this feedback (let me know), but...

I feel that the [Web Thing Model](http://model.webofthings.io/) member submission (and the [Building the Web of Things book](https://www.manning.com/books/building-the-web-of-things)) deals with these issues quite elegantly. This model proposes three integration patterns:
* **Direct** - The WoT API is exposed by the Thing itself. Useful for relatively high powered devices which support HTTP and TCP/IP and can therefore expose a REST API directly (e.g. a WiFi camera).
* **Gateway** - The WoT API is exposed by a gateway. Useful for resource-constrained devices which can’t directly connect to the web themselves and so use a gateway to bridge them to the web (e.g. ZigBee or Bluetooth sensors).
* **Cloud** - The WoT API is provided by a cloud service. Useful for a large number of IoT devices over a wide geographic area which need co-ordinating centrally. They can use whatever IoT protocols they want to communicate with the cloud service on the back end. (e.g. air quality sensors using MQTT over a LoRA WAN network).

The view that I support is that WoT standardisation should focus on the data model and REST API used on the web itself and not concern itself with non-web protocols which may be used on the back end. Where the direct integration pattern is used this will be the native protocol the Thing uses (via HTTP, CoAP or HTTP/2), but the gateway and cloud integration patterns may use any number of custom protocols or formats better suited to solve specific problems.

I'm afraid I think the idea of a universal data model used by hundreds of different proprietary IoT protocols and a universal scripting API used by all programming languages is just wishful thinking. We should focus on the web layer of the Web of Things and make the non-web parts out of scope of the charter.

My hope is that a charter more rooted in the reality of how the IoT is actually being built (Google Weave, EVRYTHNG etc.) would make it easer for members like Mozilla and Google to contribute to the existing W3C effort rather than [start from scratch](https://benfrancis.github.io/wot/).

> There are some pull requests for the charter outstanding. They are still being discussed within the IG, but I expect we will come to a consensus soon on an updated proposal and then I would like to reach out to Mozilla specifically to discuss if we have addressed your concerns.

Sorry for jumping the gun :) I just want to maximise the chance that Mozilla can contribute to this effort.

BTW, this discussion is completely off-topic for this pull request. Please feel free to merge it and take this conversation elsewhere! (I'm not personally bothered about Mozilla's logo being there either).

mkovatsc commented:
Merged as discussed in WebConf 2016-12-14
